File: searching/astar/python/test_astar.py
"""astar関数のテスト（pytest）。

このファイルは同じディレクトリ内のastar.pyのみに依存しているため、
python/フォルダを単独で取り出してもこのテストは実行できる。
"""
import math

from astar import astar

_SQRT2 = math.sqrt(2)
_EPS = 1e-6


def _path_cost(path):
    """経路の総コスト（直進1・斜めsqrt(2)の合計）を計算する。"""
    cost = 0.0
    for i in range(1, len(path)):
        dr = abs(path[i][0] - path[i - 1][0])
        dc = abs(path[i][1] - path[i - 1][1])
        cost += _SQRT2 if dr == 1 and dc == 1 else 1.0
    return cost


def _is_valid_path(grid, path, start, goal):
    """経路が制約（先頭/末尾、隣接性、障害物回避、コーナーカット禁止）を満たすか検証する。"""
    if not path:
        return False
    if path[0] != start or path[-1] != goal:
        return False

    rows, cols = len(grid), len(grid[0])
    for r, c in path:
        if not (0 <= r < rows and 0 <= c < cols):
            return False
        if grid[r][c] != 0:
            return False

    for i in range(1, len(path)):
        dr = path[i][0] - path[i - 1][0]
        dc = path[i][1] - path[i - 1][1]
        if abs(dr) > 1 or abs(dc) > 1 or (dr == 0 and dc == 0):
            return False
        if dr != 0 and dc != 0:
            r0, c0 = path[i - 1]
            if grid[r0 + dr][c0] != 0 or grid[r0][c0 + dc] != 0:
                return False

    return True


def test_no_obstacles_diagonal_path():
    """障害物なしの直線（斜め）経路は4*sqrt(2)になる"""
    grid = [[0] * 5 for _ in range(5)]
    start, goal = (0, 0), (4, 4)
    path = astar(grid, start, goal)
    assert _is_valid_path(grid, path, start, goal)
    assert abs(_path_cost(path) - 4 * _SQRT2) < _EPS


def test_wall_with_gap_detour():
    """壁に迂回が必要なケース。最短コストは8.0"""
    grid = [
        [0, 0, 0, 0, 0],
        [0, 1, 1, 1, 0],
        [0, 0, 0, 0, 0],
        [0, 1, 1, 1, 0],
        [0, 0, 0, 0, 0],
    ]
    start, goal = (0, 0), (4, 4)
    path = astar(grid, start, goal)
    assert _is_valid_path(grid, path, start, goal)
    assert abs(_path_cost(path) - 8.0) < _EPS


def test_start_equals_goal():
    """startとgoalが同一の場合は[start]のみを返す"""
    grid = [[0] * 3 for _ in range(3)]
    path = astar(grid, (1, 1), (1, 1))
    assert path == [(1, 1)]


def test_unreachable_goal():
    """到達不可能な場合は空リストを返す"""
    grid = [
        [0, 0, 0, 0, 0],
        [1, 1, 1, 1, 1],
        [0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0],
    ]
    assert astar(grid, (0, 0), (4, 4)) == []


def test_corner_cut_blocked():
    """斜め移動のコーナーカットが禁止されているため到達不可能になるケース"""
    grid = [
        [0, 1, 0],
        [1, 0, 1],
        [0, 1, 0],
    ]
    assert astar(grid, (0, 0), (2, 2)) == []

File: searching/astar/matlab/test_astar.m
% test_astar.m
%
% astar関数のテストスクリプト（MATLAB/Octave両対応）。
% このフォルダを単独で取り出してそのまま実行できる（外部依存なし）。
% ヘルパー関数(pathCost/isValidPath)はscript内のローカル関数として
% 使うとMATLAB/Octaveで定義順序の制約があるため、各テストケース内に
% 検証ロジックを直接書く方式にしている。
%
% 実行方法: MATLAB/Octave上でこのスクリプトを実行する。
%   >> test_astar
% 全テストが成功すれば "All tests passed." が表示される。

sqrt2 = sqrt(2);
eps_tol = 1e-6;

% --- 障害物なしの直線（斜め）経路。最短コストは4*sqrt(2) ---
grid1 = zeros(5, 5);
startPos = [1 1];
goalPos = [5 5];
path1 = astar(grid1, startPos, goalPos);
assert(~isempty(path1));
assert(isequal(path1(1, :), startPos) && isequal(path1(end, :), goalPos));
cost1 = 0;
for i = 2:size(path1, 1)
    dr = abs(path1(i, 1) - path1(i - 1, 1));
    dc = abs(path1(i, 2) - path1(i - 1, 2));
    assert(dr <= 1 && dc <= 1 && ~(dr == 0 && dc == 0));
    if dr == 1 && dc == 1
        cost1 = cost1 + sqrt2;
    else
        cost1 = cost1 + 1;
    end
end
assert(abs(cost1 - 4 * sqrt2) < eps_tol);

% --- 壁に迂回が必要なケース。最短コストは8.0 ---
grid2 = [
    0 0 0 0 0;
    0 1 1 1 0;
    0 0 0 0 0;
    0 1 1 1 0;
    0 0 0 0 0
];
startPos = [1 1];
goalPos = [5 5];
path2 = astar(grid2, startPos, goalPos);
assert(~isempty(path2));
assert(isequal(path2(1, :), startPos) && isequal(path2(end, :), goalPos));
cost2 = 0;
for i = 2:size(path2, 1)
    r = path2(i, 1);
    c = path2(i, 2);
    assert(grid2(r, c) == 0); % 障害物を通らない
    dr = path2(i, 1) - path2(i - 1, 1);
    dc = path2(i, 2) - path2(i - 1, 2);
    assert(abs(dr) <= 1 && abs(dc) <= 1 && ~(dr == 0 && dc == 0));
    if dr ~= 0 && dc ~= 0
        r0 = path2(i - 1, 1);
        c0 = path2(i - 1, 2);
        assert(grid2(r0 + dr, c0) == 0 && grid2(r0, c0 + dc) == 0); % コーナーカット禁止
        cost2 = cost2 + sqrt2;
    else
        cost2 = cost2 + 1;
    end
end
assert(abs(cost2 - 8.0) < eps_tol);

% --- startとgoalが同一の場合 ---
grid3 = zeros(3, 3);
path3 = astar(grid3, [2 2], [2 2]);
assert(isequal(path3, [2 2]));

% --- 到達不可能なケース（完全に壁で塞がれている） ---
grid4 = [
    0 0 0 0 0;
    1 1 1 1 1;
    0 0 0 0 0;
    0 0 0 0 0;
    0 0 0 0 0
];
path4 = astar(grid4, [1 1], [5 5]);
assert(isempty(path4));

% --- コーナーカット禁止の確認（斜めの隙間を突き抜けて到達できないこと） ---
grid5 = [
    0 1 0;
    1 0 1;
    0 1 0
];
path5 = astar(grid5, [1 1], [3 3]);
assert(isempty(path5));

disp('All tests passed.');
